design-proposals/klip-50-partition-and-offset-in-ksqldb.md

```diff
@@ -0,0 +1,253 @@
+will be analogous to those of the [existing `ROWTIME` pseudocolumn](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-14-rowtime-as-pseudocolumn.md#queries).
+In summary:
+
+* `SELECT *` projections will not include `ROWPARTITION` and `ROWOFFSET`
+* Users may choose to explicitly add `ROWPARTITION` or `ROWOFFSET` in the selection,
+  which will cause ksqlDB to copy the selected pseudocolumn(s) into the output _value_
+* `ROWPARTITION` and `ROWOFFSET` will become reserved column names, and users will not
+  be allowed to use these names either as source column names or as output column aliases.
+  (Existing persistent queries that already use these names will continue to run undisrupted.
+  See ["Compatibility Implications"](#compatibility-implications) for more.)
+
+Consistent with `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will 
+also be allowed in `WHERE`, `GROUP BY`, and `PARTITION BY` clauses, though use cases
+for the latter two are probably limited. `ROWPARTITION` and `ROWOFFSET` will also be
+allowed in UDFs.
+
+### Pull Queries
+
+Today, users are able to select the `ROWTIME` pseudocolumn as part of pull queries:
+```
+SELECT *, ROWTIME FROM my_table WHERE my_key='foo';
+```
+For multiple reasons, we propose that the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns
+should diverge from this behavior. In other words, users will NOT be allowed to select
+`ROWPARTITION` or `ROWOFFSET` as part of a pull query.
+
+The first reason is that selecting `ROWPARTITION` or `ROWOFFSET` for a table
+is semantically dubious, especially for tables that are the result of other
+computation such as an aggregation over a stream. Should `ROWPARTITION` and `ROWOFFSET`
+in such a case represent the partition and offset of the message in the source stream
+that resulted in the aggregate value, or should they represent the partition and offset
+of the aggregate value in the output table? Note that timestamps don't have this
+ambiguity as the timestamp of the two records mentioned are the same.
+
+The second reason has to do with implications on state store size. ksqlDB is able to 
+support `ROWTIME` in pull queries because the state stores from which pull queries are 
+served includes the timestamp for the latest table row (for each key). In contrast, 
+message partition and offset are not saved in the state stores. As a result, supporting 
+`ROWPARTITION` and `ROWOFFSET` in pull queries would require adding this information
+to state stores. (I think there might be a way to get the partition by other means,
+from the replica queried in order to serve the pull query, but offset information
+definitely wouldn't be available.)
+
+If a user really wants to make partition and/or offset information available for pull
+queries, they can choose to explicitly include the source partition and offset as
+part of the query that materializes their table and query for partition/offset as
+value columns as part of their pull query.
```

Comment:
I think we should plan to illustrate this in the documentation with an example materializing query. Might also make sense to add the example query in the KLIP as well.

Reply:
Sure, added an example to the KLIP. Out of curiosity, what use cases do you envision for including partition/offset as part of a pull query? It seems less natural to be interested in partition and offset for table records compared to messages in a stream.

Reply:
Yeah, I agree it's less natural. I could see it being educationally useful for folks checking their understanding about how streams and tables relate, for example to see how the aggregate row corresponds to the source stream's partition and offset. Thanks for adding it.

Reply:
partition offset for pull queries would be super useful for debugging purposes (answering "why is a certain piece of data there?")